src/secur3dit/crypto/AES256CTR.java

```diff
@@ -0,0 +1,133 @@
+        }
+
+        char[] nonceChar = new char[16];
+        for (int i = 0; i < nonceChar.length; ++i) {
+            nonceChar[i] = (char) Byte.toUnsignedInt(nonceByte[i]);
+        }
+
+        this.key = keyChar;
+        this.nonce = nonceChar;
+    }
+
+    /**
+     * This method performs bitwise-XOR on two 128-bit long data.
+     * This is required to generate the cyphertext (Output_of_AES256 XOR Plaintext)
+     * @param img       The image file loaded in char[].
+     * @param state     The encrypted state matrix char[16] from AES-256.
+     * @param blockNum  The block number which is being XORed.
+     * @return Nothing  The specific block in that image gets encrypted.
+     */
+    private static void XOR(char[] img, final char[] state, final int blockNum) {
+        final int blockStartIndex = state.length * blockNum;
+        for (int i = blockStartIndex; i < blockStartIndex + state.length && i < img.length; ++i) {
+                img[i] = (char) (img[i] ^ state[i % state.length]);
+        }
+    }
+
+    /**
+     * This method performs bitwise-XOR on two 128-bit long data.
+     * This is required to generate input to AES256 (Nonce XOR BlockNumber)
+     * @param nonce     First 128-bit data of type char[16].
+     * @param blockNum  Second 128-bit data of type char[16].
+     * @return          The output string after performing bitwise-XOR
+     */
+    private static char[] XOR(char[] nonce, int blockNum) {
+        ByteBuffer bf = ByteBuffer.allocate(nonce.length);
+        bf.order(ByteOrder.BIG_ENDIAN);
+        bf.putInt(blockNum);
+        byte[] b = bf.array();
+
+        char[] result = new char[nonce.length];
+        for (int i = 0; i < nonce.length; ++i) {
+            result[i] = (char) (nonce[i] ^ (char) Byte.toUnsignedInt(b[i]));
+        }
+        return result;
+    }
+
+    /**
+     * This method performs encryption on the image.
+     * @param ob        Instance of AES256CTR which stores the nonce and key.
+     * @param filePath  The path to the file which needs to be encrypted.
+     * @return Nothing  The file specified by {@code filePath} gets encrypted.
+     * @throws IOException
+     */
+    public static void encrypt(AES256CTR ob, String filePath) throws IOException {
+
+        // Read the file into bytes[], convert into char[]
+        byte[] bytes = Files.readAllBytes(Paths.get(filePath));
+
+        char[] img = new char[bytes.length];
+        for (int i = 0; i < img.length; ++i) {
+            img[i] = (char) Byte.toUnsignedInt(bytes[i]);
+        }
+
+        final int NUMBER_OF_BLOCKS = img.length / 16 + (img.length % 16 == 0 ? 0 : 1);
+        for (int i = 0; i < NUMBER_OF_BLOCKS; ++i) {
+            char[] result = XOR(ob.nonce, i);
+            AES256.encrypt(result, ob.key);
+            XOR(img, result, i);
+        }
+
+        for (int i = 0; i < img.length; ++i) {
+            bytes[i] = (byte) img[i];
+        }
+        Files.write(Paths.get(filePath), bytes);
```

Comment:
I think we should keep the generated output file separate from the input file. This can help in two ways:
1. Provide the user, control over the original file.
2. Preserve the bytes of the original encrypted file upon an incorrect decryption attempt.

It could be debated that this lead to the usage of twice the space in the disk. To tackle this, we could provide an option in the user interface which would let the user delete the original file. 

Let me know if you think otherwise.

Reply:
This makes absolute sense. However, to keep the codebase as modular as possible, the encryption/decryption have to work on a particular file. The creation of a copy, however, can be (and should be) done at the UI end where it would make more sense.

Tagging @namannihal as we would need to deal with this.
PS: This approach could also be followed for the filters.

Reply:
All right, that works too!